// @tailwindcss-mangle/config/dist/index.mjs
import path from 'node:path';
import fs from 'node:fs/promises';
import { createDefineConfig, loadConfig } from 'c12';
import process from 'node:process';
import { defaultMangleClassFilter } from '@tailwindcss-mangle/shared';
import { CSS_LANGS_RE } from 'is-css-request';

const defaultPipelineInclude = ["**/*.{html,js,ts,jsx,tsx,vue,svelte,astro,elm,php,phtml,mdx,md}", CSS_LANGS_RE];
const defaultPipelineExclude = [/[\\/](node_modules|dist|\.temp|\.cache|\.vscode)[\\/]/];
function getDefaultPatchConfig() {
  return {
    output: {
      filename: ".tw-patch/tw-class-list.json",
      removeUniversalSelector: true,
      loose: true
    },
    tailwindcss: {}
  };
}
function getDefaultMangleUserConfig() {
  return {
    mangleClassFilter: defaultMangleClassFilter,
    include: defaultPipelineInclude,
    exclude: defaultPipelineExclude,
    disabled: process.env.NODE_ENV === "development",
    classListPath: ".tw-patch/tw-class-list.json",
    classMapOutput: {
      enable: false,
      filename: ".tw-patch/tw-map-list.json",
      loose: true
    },
    preserveFunction: []
  };
}
function getDefaultUserConfig() {
  return {
    patch: getDefaultPatchConfig(),
    mangle: getDefaultMangleUserConfig()
  };
}

const configName = "tailwindcss-mangle";

function getConfig(cwd) {
  return loadConfig({
    name: configName,
    defaults: {
      ...getDefaultUserConfig()
    },
    cwd
  });
}
const defineConfig = createDefineConfig();
function initConfig(cwd) {
  return fs.writeFile(
    path.resolve(cwd, `${configName}.config.ts`),
    `import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({})
`,
    "utf8"
  );
}

export { configName, defineConfig, getConfig, getDefaultMangleUserConfig, getDefaultPatchConfig, getDefaultUserConfig, initConfig };
